integrations/wagmi-v2/src/index.ts

```diff
@@ -63,3 +84,141 @@ export function injectedWithSapphire(
+ *
+ * ```typescript
+ *    import { createSapphireConfig } from '@oasisprotocol/sapphire-wagmi-v2';
+ *
+ *    export const config = createSapphireConfig({
+ *    	sapphireConfig: {
+ *          replaceProviders: true
+ *    	}
+ *      ...
+ *    });
+ * ```
+ *
+ * In case your dApp is multichain, the bellow example will create duplicate connectors for Sapphire:
+ *
+ * ```typescript
+ *    import { createSapphireConfig } from '@oasisprotocol/sapphire-wagmi-v2';
+ *
+ *    export const config = createSapphireConfig({
+ *      sapphireConfig: {
+ *        replaceProviders: false,
+ *        // Define which providers you want to wrap via RDNS
+ *        wrappedProvidersFilter: (rdns) => ['io.metamask'].includes(rdns)
+ *      }
+ *      ...
+ *    });
+ * ```
+ *
+ * @param {CreateConfigParameters<chains, transports, connectorFns> & CreateSapphireConfigParameters} parameters - Extended wagmi parameters,
+ * with sapphireConfig
+ * @return {Config<chains, transports, connectorFns>} Wagmi config
+ */
+export function createSapphireConfig<
+	chains extends readonly [Chain, ...Chain[]],
+	transports extends Record<chains[number]["id"], Transport>,
+	connectorFns extends readonly CreateConnectorFn[],
+>(
+	parameters: CreateConfigParameters<chains, transports, connectorFns> &
+		CreateSapphireConfigParameters,
+): Config<chains, transports, connectorFns> {
+	const { sapphireConfig, ...restParameters } = parameters;
+	const {
+		replaceProviders = false,
+		wrappedProvidersFilter = () => true,
+		wrap: wrapOptions,
+	} = sapphireConfig;
+
+	const _addEventListener = EventTarget.prototype.addEventListener;
+	Object.defineProperty(EventTarget.prototype, "addEventListener", {
+		value: function (
+			type: string,
+			callback: EventListenerOrEventListenerObject | null,
+			options?: AddEventListenerOptions | boolean,
+		): void {
+			if (type === EIP6963_ANNOUNCE_PROVIDER_EVENT_NAME) {
+				_addEventListener.call(
+					this,
+					type,
+					(event: Event) => {
+						let patchCustomEvent = null;
+
+						if (
+							SUPPORTED_RDNS.includes(
+								(event as EIP6963AnnounceProviderEvent).detail.info.rdns,
+							)
+						) {
+							const announceProviderEvent =
+								event as EIP6963AnnounceProviderEvent;
+
+							if (replaceProviders) {
+								if (
+									!isWrappedEthereumProvider(
+										announceProviderEvent.detail.provider,
+									)
+								) {
+									patchCustomEvent = new CustomEvent(
+										EIP6963_ANNOUNCE_PROVIDER_EVENT_NAME,
+										{
+											detail: {
+												...announceProviderEvent.detail,
+												provider: wrapEthereumProvider(
+													announceProviderEvent.detail.provider,
+													wrapOptions,
+												),
+											},
+										},
+									);
+								}
+							} else if (
+								wrappedProvidersFilter(
+									announceProviderEvent.detail.info.rdns,
+								) &&
+								!isWrappedEthereumProvider(
+									announceProviderEvent.detail.provider,
+								)
```

Comment:
(optional) Maybe name and lift this up earlier 🤔

Reply:
Like so https://github.com/oasisprotocol/sapphire-paratime/pull/550/commits/720f2524f8ee4bfac6ead7164cea26a8c00fb256?

Reply:
Works for me :)

I was thinking just something basic would do too.

```ts
condition = wrappedProvidersFilter(announceProviderEvent.detail.info.rdns,) && !isWrappedEthereumProvider(announceProviderEvent.detail.provider)
```

Reply:
I am not in favor of extracting if statement/s condition to a variable, it decreases readability IMO. Maybe in cases where it would speed up execution.